ledger-core/v2/conveyor/smachine/state_updates.go

```diff
@@ -159,26 +154,35 @@ func init() {
 			filter: updCtxExec,
 			params: updParamStep | updParamUint,
 
-			shortLoop: func(slot *Slot, stateUpdate StateUpdate, loopCount uint32) bool {
+			shortLoop: func(slot *Slot, stateUpdate StateUpdate, loopCount uint32, nextDecl *StepDeclaration) bool {
 				if loopCount >= stateUpdate.param0 {
 					return false
 				}
-				switch nextStep := stateUpdate.step.Transition; {
-				case nextStep == nil:
+
+				nextStep := stateUpdate.step.Transition
+				if nextStep == nil {
 					slot.setNextStep(stateUpdate.step, nil)
 					return false // the same step won't be short-looped
+				}
 
-				case slot.stepDecl != nil:
-					prevSeqID := slot.stepDecl.SeqID
-					nextDecl := slot.declaration.GetStepDeclaration(nextStep)
-					slot.setNextStep(stateUpdate.step, nextDecl)
-					return nextDecl != nil && prevSeqID < nextDecl.SeqID // only proper further steps can be short-looped
+				if nextDecl == nil {
+					nextDecl = slot.declaration.GetStepDeclaration(nextStep)
+				}
+
+				curStep := slot.step.Transition
+				prevSeqID := 0
+				if slot.stepDecl != nil {
+					prevSeqID = slot.stepDecl.SeqID
+				}
+				slot.setNextStep(stateUpdate.step, nextDecl)
 
-				default:
-					isConsec, nextDecl := slot.declaration.IsConsecutive(slot.step.Transition, nextStep)
-					slot.setNextStep(stateUpdate.step, nextDecl)
-					return isConsec
+				if nextDecl != nil && prevSeqID != 0 {
+					if nextSeqID := nextDecl.SeqID; nextSeqID != 0 {
+						return nextSeqID > prevSeqID // only proper forward steps can be short-looped
+					}
 				}
+				isConsecutive := slot.declaration.IsConsecutive(curStep, nextStep)
+				return isConsecutive
```

Comment:
FYI - it doesn't seem that there is a single unit test for this piece of code. You are making the changes blindly, there is no way to be sure that nothing was broken.

Reply:
well, there is a threshold for me - when algo is complex, then I do very elaborated tests, e.g. like in merkler. Here the complexity is relatively low, so I'd rather delegate it to someone else.